File: Solutions/Python/combination_sum.py
"""
    [MEDIUM]
    39. Combination Sum

    Concepts:
    - backtracking

    Stats:
        Runtime | 39 ms     [Beats 9.92%]
        Memory  | 17.98 MB  [Beats 32.81%]
"""

class Solution:
    def combinationSum(self, cands: List[int], target: int) -> List[List[int]]:
        combos = []
        cands.sort()

        def dfs(idx: int, path: list[int], curr: int):
            nonlocal combos

            if curr == target:
                combos.append(path)

            for i, num in enumerate(cands[idx:]):
                if curr < target:
                    dfs(i + idx, path + [num], curr + num)

        dfs(0, [], 0)

        return combos

File: Solutions/Cpp/combination_sum.cpp
/*
    [MEDIUM]
    39. Combination Sum

    Concepts:
    - backtracking

    Stats:
        Runtime | 4 ms      [Beats 43.79%]
        Memory  | 15.21 MB  [Beats 38.85%]
*/

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& cands, int target) {
        std::vector<std::vector<int>> combos = {};
        std::vector<int> path = {};
        std::sort(cands.begin(), cands.end());

        std::function<void(int, int)> dfs;
        dfs = [&](int idx, int curr){
            if (curr == target)
                combos.push_back(path);
            else if (curr > target)
                return;

            for (int i = idx; i < cands.size(); i++) {
                if (curr < target) {
                    path.push_back(cands[i]);
                    dfs(i, curr + cands[i]);
                    path.pop_back();
                }
            }
        };

        dfs(0, 0);
        return combos;
    }
};